### ISSUE TITLE
Adaptation to other OAuth2 services: Pull requests or forks?

### ISSUE THREAD

**Issue by mbland** [bitly/google_auth_proxy#65]
I've managed to successfully [fork this repo and adapt it to a new OAuth2 provider](https://github.com/bitly/google_auth_proxy/compare/master...18F:myusa-staging). I know others have done the same. However, given how little work it was, I'm wondering if you'd be open [to generalizing the server to support multiple providers out-of-the-box](https://github.com/bitly/google_auth_proxy/compare/master...18F:generalize), defaulting to Google but allowing for other providers.

I'm happy to take this on, and add tests as I go, as you can see from the latter link. (I'm about to file a PR for a bug I found in `Options.Validate` as part of this process.) That said, I understand if you'd rather keep this instance "pure" and let forks for other providers proliferate as they will.

Apologies if you've address this before; didn't see this come up in any earlier issue. Also, thanks for writing this; we've been using it successfully for months now on our Google-managed domain.

**Comment by jehiah** [comment]
@mbland good questions. Let me dig into your current set of changes to understand what would be involved.

I'm not opposed to some generalization for other oauth provider endpoints and requested scopes, but i suspect it will become tricky as oauth providers don't have a consistent way to echo back identity after authentication which GAP needs.

I'm also keen to eventually support #28 which likely ties some functionality closer to Google's apis.

**Comment by jehiah** [comment]
also cc: @balshor w/r/t 5f8bd41a7c089fdf047380acc9b02f2c19605721

**Comment by mbland** [comment]
OK, I'll keep hacking on [18F/oauth2_proxy](https://github.com/18F/oauth2_proxy) to see if I can come up with some seams that make sense, and we can consider later whether they're worth pushing back up here. Feel free to subscribe to our repo, jump in on PRs, etc.

**Comment by mbland** [comment]
FYI, I've merged the latest changes from here into [the `generalize` branch of our oauth2_proxy fork](https://github.com/bitly/google_auth_proxy/compare/master...18F:generalize?expand=1), and we're running this version now. Seems to work well.

If you're curious, I created a `providers` package to present a unified `providers.Provider` interface between the Google and MyUSA providers, and an `api` package to share the common `api.Request()` (formerly `apiRequest()`) between them.

I still need to add proper tests, but in the off chance you'd be interested in having any of this pushed upstream, let me know and I can prioritize that.
